src/main/scala/com/horizon/exchangeapi/BusinessRoutes.scala

```diff
@@ -864,68 +872,174 @@ trait BusinessRoutes extends JacksonSupport with AuthenticationSupport {
+    val compositeId: String = OrgAndId(orgid, policy).toString
+    exchAuth(TNode(OrgAndId(orgid, "*").toString), Access.READ) { ident =>
+      validateWithMsg(if(!((!(reqBody.changedSince < 0L)) && (reqBody.numEntries.isEmpty || !(reqBody.numEntries.getOrElse(-1) < 0)))) Some(ExchMsg.translate("bad.input")) else None) {
         complete({
-          val nodeOrgids = reqBody.nodeOrgids.getOrElse(List(orgid)).toSet
+          val nodeOrgids: Set[String] = reqBody.nodeOrgids.getOrElse(List(orgid)).toSet
           var searchSvcUrl = ""    // a composite value (org/url), will be set later in the db.run()
           db.run(BusinessPoliciesTQ.getService(compositeId).result.flatMap({ list =>
-            logger.debug("POST /orgs/" + orgid + "/business/policies/" + policy + "/search getService size: " + list.size)
             if (list.nonEmpty) {
               // Finding the service was successful, form the query for the nodes for the next step
-              val service = BusinessPoliciesTQ.getServiceFromString(list.head)    // we should have found only 1 business pol service string, now parse it to get service list
+              val service: BService = BusinessPoliciesTQ.getServiceFromString(list.head)    // we should have found only 1 business pol service string, now parse it to get service list
+              val optArch: Option[String] =
+                if(service.arch.equals("") || service.arch.equals("*"))
+                  None
+                else
+                  Some(service.arch)
+              
               searchSvcUrl = OrgAndId(service.org, service.name).toString
+              
               /*
                 Narrow down the db query results as much as possible by joining the Nodes and NodeAgreements tables and filtering.
                 In english, the join gets: n.id, n.nodeType, n.publicKey, a.serviceUrl, a.state
                 The filters are: n is in the given list of node orgs, n.pattern is not set, the node is not stale, the node arch matches the service arch (the filter a.state=="" is applied later in our code below)
                 After this we have to go thru all of the results and find nodes that do NOT have an agreement for searchSvcUrl.
                 Note about Slick usage: joinLeft returns node rows even if they don't have any agreements (which is why the agreement cols are Option() )
               */
-              val oldestTime = if (reqBody.changedSince > 0) ApiTime.thenUTC(reqBody.changedSince) else ApiTime.beginningUTC
-              val nodeQuery = for {
-                (n, a) <- NodesTQ.rows.filter(_.orgid inSet(nodeOrgids)).filter(_.pattern === "").filter(_.publicKey =!= "").filter(_.lastUpdated >= oldestTime).filter(n => {n.arch === service.arch || service.arch == "" || service.arch == "*"}) joinLeft NodeAgreementsTQ.rows on (_.id === _.nodeId)
-              } yield (n.id, n.nodeType, n.publicKey, a.map(_.agrSvcUrl), a.map(_.state))
+              val pagination =
```

Comment:
Can you add some comments before each block in the query just detailing what it does?